File: [74]搜索二维矩阵.cpp
#include <vector>
using namespace std;

class Solution {
public:
	bool searchMatrix(vector<vector<int>>& matrix, int target) {
		if (matrix.empty() || matrix[0].empty())
			return false;

		int left = 0, right = matrix.size();
		while (left < right) {
			int mid = left + (right - left) / 2;
			if (matrix[mid][0] == target)
				return true;
			else if (matrix[mid][0] < target)
				left = mid + 1;
			else
				right = mid;
		}
		if (left == 0)
			return false;
		int row = left - 1;
		left = 0, right = matrix[0].size();
		while (left < right) {
			int mid = left + (right - left) / 2;
			if (matrix[row][mid] == target)
				return true;
			else if (matrix[row][mid] < target)
				left = mid + 1;
			else
				right = mid;
		}
		return false;
	}
};

File: solutions/[74]搜索二维矩阵.py
from typing import List


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        def binary_search(arr):
            l, r = 0, len(arr)
            while l < r:
                mid = (l+r) // 2
                if arr[mid] > target:
                    r = mid
                else:
                    l = mid + 1
            return l - 1
        
        v_p = binary_search([line[0] for line in matrix])
        l_p = binary_search(matrix[v_p])

        # print(v_p, l_p, matrix[v_p][l_p])
        return matrix[v_p][l_p] == target
